### PR TITLE
Python: Adding MongoDB Atlas Vector Search Connector

### PR DESCRIPTION
**Pull request by Jibola**
### Motivation and Context
Resolves: #2591 

### Description

* Added in the `MongoDBAtlasMemoryStore`. A memory store abstraction that facilitates connections between a MongoDB Atlas cluster to conduct [Atlas Vector Search](https://www.mongodb.com/products/platform/atlas-vector-search) on the Microsoft Semantic Kernel.

* Leverages our async python driver [motor](https://motor.readthedocs.io/en/stable/) to keep with the heuristic of asynchronous code

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### Callout
* When writing my test, I noticed there's a habit of other tests conducting a `try` on imports. Grepping through those implementations, I see some of them [encapsulate their respective libraries in function calls rather than doing direct imports in the module.](https://github.com/microsoft/semantic-kernel/blob/main/python/semantic_kernel/connectors/memory/chroma/chroma_memory_store.py#L55-L63) Is this the general code pattern we should follow? Or is it fine to have our imports at the top of the module fine?

### CODE REVIEW

**Comment by Jibola** [comment]
> @Jibola please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement



@microsoft-github-policy-service agree company="MongoDB"

File: python/semantic_kernel/connectors/memory/mongodb_atlas_vector_search/atlas_vector_search_memory_store.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+            List[str] -- The unique identifiers for the memory records.
+        """
+        await self.create_collection_async(collection_name)
+
+        upserts: List[UpdateOne] = []
+        for record in records:
+            if not record._id:
+                record._id = str(uuid.uuid4())
+                self._logger.debug(
+                    "record id not found for Memory Store, mutated with new uuid {}".format(
```
Comment (blink1073) [inline python/semantic_kernel/connectors/memory/mongodb_atlas_vector_search/atlas_vector_search_memory_store.py:190]: The Python [docs](https://docs.python.org/3/library/logging.html#logging.debug) recommend passing the formatting args as `*args`.

File: python/semantic_kernel/utils/settings.py [hunk 2]
```diff
@@ -168,6 +168,20 @@ def bing_search_settings_from_dot_env() -> str:
     return api_key
 
 
+def mongodb_atlas_settings_from_dot_env() -> str:
+    """Returns the Atlas MongoDB Connection String from the .env file
```
Comment (blink1073) [inline python/semantic_kernel/utils/settings.py:172]:
```suggestion
    """Returns the Atlas MongoDB Connection String from the .env file.
```

**Review by blink1073 — COMMENTED** [review]
Looking good so far!  I see we're waiting on a new credential and there is one TODO.  Otherwise I just saw a few nits.

File: python/semantic_kernel/connectors/memory/mongodb_atlas_vector_search/atlas_vector_search_memory_store.py (hunk 1 above)
Comment (Jibola) [reply python/semantic_kernel/connectors/memory/mongodb_atlas_vector_search/atlas_vector_search_memory_store.py:190]: Yeah, agreed. However, it looks like the `NullLogger` used here doesn't marry well with that approach. 
```
>       self._logger.debug(
            "matched_count=%s, upserted_count=%s",
            bulk_update_result.matched_count,
            bulk_update_result.upserted_count,
        )
E       TypeError: debug() takes 2 positional arguments but 4 were given
```

I went ahead and looked at the [`NullLogger`](https://github.com/microsoft/semantic-kernel/blob/main/python/semantic_kernel/utils/null_logger.py) implementation and now think it makes sense to have a small PR to change the function signatures of the `NullLogger` to just match the base `Logger` class. 

cc: @awharrison-28 @shawncal for thoughts?

File: python/tests/integration/connectors/memory/test_atlas_vector_search.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        try:
+            yield memory
+        except errors.OperationFailure as e:
+            # DuplicateIndex failures in test are due to lag in index deletions
+            if e.code != DUPLICATE_INDEX_ERR_CODE:
+                raise
+        finally:
+            for cname in await memory.get_collections_async():
+                await memory.delete_collection_async(cname)
+            # TODO Possibly remove
```
Comment (blink1073) [inline python/tests/integration/connectors/memory/test_atlas_vector_search.py:82]: Flagging this so we don't forget about it.

Comment (Jibola) [reply python/tests/integration/connectors/memory/test_atlas_vector_search.py]: Thanks for flagging! I managed to place a much smarter mitigation in its place!

**Comment by Jibola** [comment]
@microsoft-github-policy-service agree company="MongoDB"

File: python/semantic_kernel/connectors/memory/mongodb_atlas_vector_search/atlas_vector_search_memory_store.py (hunk 1 above)
Comment (awharrison-28) [reply python/semantic_kernel/connectors/memory/mongodb_atlas_vector_search/atlas_vector_search_memory_store.py]: That makes sense. @Jibola can you create an issue to track this work?

File: python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    async def create_collection_async(self, collection_name: str) -> None:
+        """Creates a new collection in the data store.
+
+        Arguments:
+            collection_name {str} -- The name associated with a collection of embeddings.
+
+        Returns:
+            None
+        """
+        # Defining the search index enforces its creation
```
Comment (Jibola) [inline python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py:78]:
```suggestion
```

File: python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+            collection_name {str} -- The name associated with a collection of embeddings.
+            record {MemoryRecord} -- The memory record to upsert.
+
+        Returns:
+            str -- The unique identifier for the memory record.
+        """
+        await self.create_collection_async(collection_name)
+
+        if not record._id:
+            record._id = str(uuid.uuid4())
```
Comment (awharrison-28) [inline python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py:137]: @Jibola does the ID of a mongoDB vector record have to a UUID? 

Ideally, the unique ID for a vector used by SK is a natural language string (see python/samples/kernel-syntax-examples/memory.py). We realize that some vector stores don't support this though, and we've had to implement some interesting logic to get around this - see qdrant_memory_store.py. Any memory connector should be able to seemlessly replace VolatileMemoryStore - with the exception of index creation which may have to be done outside of SK.

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py (hunk 5 above)
Comment (Jibola) [reply python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py:134]: No, it doesn't need to be a uuid at all. I had been working off assumptions seen from other implementations. My question then is, _is record._id guaranteed?_. 
* If yes, then I can remove this section outright

* If not, then my next question would be _does it have to be unique?_
  * If yes, then we'd have to make _something_ the string representation or throw an exception so the end-user knows
  * If not, then we can demote id from being the primary key in the document, and populate and arbitrary--somewhat untracked--ObjectId to be the "primary key" and would be obfuscated from code. 

Let me know what version works best here

Comment (awharrison-28) [reply python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py]:
```
class MemoryRecord:
    _key: str
    _timestamp: Optional[datetime]
    _is_reference: bool
    _external_source_name: Optional[str]
    _id: str
    _description: Optional[str]
    _text: Optional[str]
    _additional_metadata: Optional[str]
    _embedding: ndarray
```

Pasting the attributes of MemoryRecord above for reference. 

Here are a list of scenarios and the expectations for each: 

1. The data is completely new (insert):
**In this scenario, MemoryRecord._key will be empty. When creating the mongoDB record, MemoryRecord._key = MemoryRecord._id. For vector stores that do not support natural language keys, MemoryRecord._key != MemoryRecord._id, but it sounds like we don't have to worry about that for Atlas :)**

2. The data is not new (intentional update):
**In this scenario, MemoryRecord._key will be populated. Upsert will act as an update.**

3. The data is intended to be new, but there's a key collision (unintentional update):
**This scenario is identical to scenario 2, but unintentional. We recognize that key collisions are more likely to happen when using natural language as keys. We intentionally called this part of SK 'semantic memory' to associate what it stores with relevance as opposed to a dumping ground of information. Ensuring that keys can be searchable and 'relevant' is an important aspect of our vision for SK. It does come with different assumptions than traditional big data storage, and it It is up to the app layer to work with those assumptions.**

Comment (awharrison-28) [reply python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py]: cool thank you!

File: python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+            If the record does not exist, it will be created.
+
+        Arguments:
+            collection_name {str} -- The name associated with a collection of embeddings.
+            records {MemoryRecord} -- The memory records to upsert.
+
+        Returns:
+            List[str] -- The unique identifiers for the memory records.
+        """
+        await self.create_collection_async(collection_name)
```
Comment (awharrison-28) [inline python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py:155]: semantic_text_memory already calls this upon save_info or save_reference. This call is redundant when used with SK.

File: python/semantic_kernel/connectors/memory/mongodb_atlas/utils.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    Arguments:
+        data {dict} -- Azure Cognitive Search result data.
+
+    Returns:
+        MemoryRecord -- The MemoryRecord from Azure Cognitive Search Data Result.
+    """
+    meta = data.get(MONGODB_FIELD_METADATA, {})
+
+    return MemoryRecord(
+        id=meta.get(MONGODB_FIELD_ID),
```
Comment (awharrison-28) [inline python/semantic_kernel/connectors/memory/mongodb_atlas/utils.py:35]: I believe you also want key=meta.get(MONGODB_FIELD_ID)

**Review by awharrison-28 — CHANGES_REQUESTED** [review]
just 2 small nits

**Review by awharrison-28 — APPROVED** [review]

**Review by shawncal — APPROVED** [review]

**Review by deehoc — APPROVED** [review]

**Review by alexchaomander — APPROVED** [review]
Thank you for adding this connector!

**Review by awharrison-28 — APPROVED** [review]

**Review by lemillermicrosoft — APPROVED** [review]

**Pull request merged**